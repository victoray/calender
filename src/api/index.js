import {useDispatch, useStore} from "react-redux";
import {useEffect, useState} from "react";
import axios from "axios";

/**
 * useGetHolidays custom Hook, makes a request to the API and retrieves the holidays from the give range.
 * @params: {Date} [startDate]
 * @params: {Date} [endDate]
 * @return: Object
 */
export const useGetHolidays = (startDate, endDate) => {
    const state = useStore().getState();
    const [holidays, setHolidays] = useState(state && state.holidays);
    const dispatch = useDispatch();

    useEffect(() => {
        // Sort the list of already retrieved holidays
        let holidayList = [];
        if (state.holidays) {
            holidayList = [...Object.keys(state.holidays)];
        }
        holidayList.sort((a, b) => new Date(a) - new Date(b));

        const start = holidayList[0];
        const end = holidayList[holidayList.length - 1];


        // if the current date range is in the redux store return it else make new request
        if ((new Date(start) <= startDate && endDate <= new Date(end))) {
            setHolidays(state.holidays);
        } else {
            axios.post("https://wozmx9dh26.execute-api.eu-west-1.amazonaws.com/api/holidays", {
                apiKey: "887d33535711e966f7b5746f263b4b81",
                startDate: startDate.toISOString().substring(0, 10),
                endDate: endDate.toISOString().substring(0, 10)
            }).then(r => {
                dispatch({type: "GET_HOLIDAYS", payload: r.data.holidays});
                setHolidays(r.data.holidays);
            }).catch(e => console.log(e))
        }

        // eslint-disable-next-line
    }, [startDate, endDate, dispatch]);

    return holidays;
};
