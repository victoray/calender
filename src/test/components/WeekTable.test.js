import React from "react";
import {mount, shallow} from "enzyme";
import WeekTableData, {WeekTable} from "../../components/WeekTable";
import {useGetHolidays} from "../../api";

jest.mock('../../api/');

const response = {
    error: false,
    holidays: {
        "2019-02-02": [
            {
                name: "Küünlapäev ehk pudrupäev",
                "type": "folk"
            }
        ],
        "2019-02-09": [
            {
                "name": "Luuvalupäev",
                "type": "folk"
            }
        ],
        "2019-02-22": [
            {
                "name": "Talvine peetripäev",
                "type": "folk"
            }
        ],
        "2019-02-24": [
            {
                "name": "Iseseisvuspäev",
                "type": "public"
            },
            {
                "name": "Talvine madisepäev",
                "type": "folk"
            }
        ]
    }
};

it('should render <WeekTable/>', () => {

    useGetHolidays.mockReturnValue({
        data: response
    });
    const date = new Date("2019-02-02");
    const component = mount(<WeekTable startDate={date} holidays={response.holidays}/>);

    expect(component).toMatchSnapshot();

});

it('should render <WeekTableData/> and call useGetHolidays', () => {

    useGetHolidays.mockReturnValue({
        data: response
    });
    const date = new Date("2019-02-02");
    const startDate = new Date(date);
    startDate.setDate(startDate.getDate() - 7);
    const component = shallow(<WeekTableData startDate={startDate} endDate={date}/>);

    expect(useGetHolidays).toBeCalledWith(startDate, date);
    expect(component).toMatchSnapshot();

});