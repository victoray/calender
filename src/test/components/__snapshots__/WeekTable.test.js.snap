// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <WeekTable/> 1`] = `
<WeekTable
  holidays={
    Object {
      "2019-02-02": Array [
        Object {
          "name": "Küünlapäev ehk pudrupäev",
          "type": "folk",
        },
      ],
      "2019-02-09": Array [
        Object {
          "name": "Luuvalupäev",
          "type": "folk",
        },
      ],
      "2019-02-22": Array [
        Object {
          "name": "Talvine peetripäev",
          "type": "folk",
        },
      ],
      "2019-02-24": Array [
        Object {
          "name": "Iseseisvuspäev",
          "type": "public",
        },
        Object {
          "name": "Talvine madisepäev",
          "type": "folk",
        },
      ],
    }
  }
  startDate={2019-02-02T00:00:00.000Z}
>
  <div
    className="scrollable"
  >
    <Table
      as="table"
      celled={true}
      columns={7}
      selectable={true}
      striped={true}
      unstackable={true}
    >
      <table
        className="ui celled selectable striped unstackable seven column table"
      >
        <TableHeader
          as="thead"
        >
          <thead
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableHeaderCell
                  as="th"
                  key="0"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Sat
                      </p>
                      <strong>
                        2.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="1"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Sun
                      </p>
                      <strong>
                        3.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="2"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Mon
                      </p>
                      <strong>
                        4.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="3"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Tue
                      </p>
                      <strong>
                        5.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="4"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Wed
                      </p>
                      <strong>
                        6.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="5"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Thu
                      </p>
                      <strong>
                        7.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
                <TableHeaderCell
                  as="th"
                  key="6"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  <TableCell
                    as="th"
                    className=""
                    style={
                      Object {
                        "textAlign": "center",
                      }
                    }
                  >
                    <th
                      className=""
                      style={
                        Object {
                          "textAlign": "center",
                        }
                      }
                    >
                      <p>
                        Fri
                      </p>
                      <strong>
                        8.2.2019
                      </strong>
                    </th>
                  </TableCell>
                </TableHeaderCell>
              </tr>
            </TableRow>
          </thead>
        </TableHeader>
        <TableBody
          as="tbody"
        >
          <tbody
            className=""
          >
            <TableRow
              as="tr"
              cellAs="td"
            >
              <tr
                className=""
              >
                <TableCell
                  as="td"
                  key="0"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  >
                    <Message
                      info={true}
                      key="0"
                      success={false}
                    >
                      <div
                        className="ui info message"
                      >
                        <p>
                          Küünlapäev ehk pudrupäev
                        </p>
                      </div>
                    </Message>
                  </td>
                </TableCell>
                <TableCell
                  as="td"
                  key="1"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
                <TableCell
                  as="td"
                  key="2"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
                <TableCell
                  as="td"
                  key="3"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
                <TableCell
                  as="td"
                  key="4"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
                <TableCell
                  as="td"
                  key="5"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
                <TableCell
                  as="td"
                  key="6"
                  style={
                    Object {
                      "height": 50,
                    }
                  }
                  verticalAlign="top"
                >
                  <td
                    className="top aligned"
                    style={
                      Object {
                        "height": 50,
                      }
                    }
                  />
                </TableCell>
              </tr>
            </TableRow>
          </tbody>
        </TableBody>
      </table>
    </Table>
  </div>
</WeekTable>
`;

exports[`should render <WeekTableData/> and call useGetHolidays 1`] = `
<WeekTable
  holidays={
    Object {
      "data": Object {
        "error": false,
        "holidays": Object {
          "2019-02-02": Array [
            Object {
              "name": "Küünlapäev ehk pudrupäev",
              "type": "folk",
            },
          ],
          "2019-02-09": Array [
            Object {
              "name": "Luuvalupäev",
              "type": "folk",
            },
          ],
          "2019-02-22": Array [
            Object {
              "name": "Talvine peetripäev",
              "type": "folk",
            },
          ],
          "2019-02-24": Array [
            Object {
              "name": "Iseseisvuspäev",
              "type": "public",
            },
            Object {
              "name": "Talvine madisepäev",
              "type": "folk",
            },
          ],
        },
      },
    }
  }
  startDate={2019-01-26T00:00:00.000Z}
/>
`;
